### PR TITLE
Replace StringBuilder with Fabulous.AST

### PR DESCRIPTION
**Pull request by nojaf**
Hello, I saw this project on my Twitter today and a part of me just dies when I see people creating F# code using StringBuilders.
The Fantomas code formatter is perfect for generating code from AST.
If you ever want to support more complex scenarios like Discriminated Unions it is really easy to use the F# syntax tree from the compiler.

I used [this site](https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWQAuAngA6wwAWVjIAroi6wAtmAC8YYAB0ATkuQQIrZMxqJmAazDY0XFSgDmYPadXrNYAG60xgk2ct3kAX0ghvQA) to figure out what AST I needed to create, and I just mapped from A to B.
There is documentation available as well at https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-syntax.html.

Feel free to do with this PR what you want. I just wanted to point out that this exists.

### CODE REVIEW

**Comment by WhiteBlackGoose** [comment]
Thanks for the PR!
> a part of me just dies when I see people creating F# code using StringBuilders

Haha, yeah, I'm not an expert in F# at all yet. Thanks, I definitely can learn from this.

But could you explain how it benefits? It heavily complicated the existing code.

> If you ever want to support more complex scenarios like Discriminated Unions it is really easy to use the F# syntax tree from the compiler.

Hm, but isn't it easy to do it with a string builder too? I mean... I admit that it may be not too formal way, but on the other hand, is there a need to complicate it, if it works well in the current shape?

**Comment by nojaf** [comment]
In its current state, it is arguably a complication. Yet you are just building a string that is very fragile and can easily introduce a bug when you add the next feature. You will know it once you ran the tool and manually inspected the output.
In this PR you can catch problems early on because you are mapping your own data structure to something more universal.

> It heavily complicated the existing code.

That is a bit subjective, yes there are many types involved which seem foreign to you.
But in essence, the structure is quite clear IMHO, you map the incoming types to a `SynTypeDefn` and then put them in an `AnonModule` inside a file.

The original `escape` function you have was far from complete and Fantomas covers things like this out of the box.

Projects like Hawaii, Snowflake and Myriad all use Fantomas and I recommend it for any use-case you might have when generating code.

**Comment by ThisFunctionalTom** [comment]
> In its current state, it is arguably a complication. Yet you are just building a string that is very fragile and can easily introduce a bug when you add the next feature. You will know it once you ran the tool and manually inspected the output. In this PR you can catch problems early on because you are mapping your own data structure to something more universal.
> 
> > It heavily complicated the existing code.
> 
> That is a bit subjective, yes there are many types involved which seem foreign to you. But in essence, the structure is quite clear IMHO, you map the incoming types to a `SynTypeDefn` and then put them in an `AnonModule` inside a file.
> 
> The original `escape` function you have was far from complete and Fantomas covers things like this out of the box.
> 
> Projects like Hawaii, Snowflake and Myriad all use Fantomas and I recommend it for any use-case you might have when generating code.

@nojaf Imagine you are new to F# and you see the string implementation and the fantomas one. Which one would you understand and which one would be possible for you to maintain. I have a few years of F# experience and I am still not able to write syntax trees with code like that. I can barely read them. There are a lot of parameters for each node and one should now the language to know the nesting of all the nodes. Is there a trick how one can make it simple? Is there somewhere a good documentation where one can learn it? Asking probably for @WhiteBlackGoose and myself :-)

**Comment by nojaf** [comment]
Being new to F# doesn't strike me as the limiting factor here.
The syntax tree is nothing more than a bunch of DU's and records.
By using this, you actually develop a deeper understanding of the language.

>  Is there a trick how one can make it simple?

Yup, there totally is, I mentioned this in my first post.
You can use the [AST viewer](https://fsprojects.github.io/fantomas-tools/#/ast) to see what [a code snippet](https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWQAuAngA6wwAWVjIAroi6wAtmAC8YYAB0ATkuQQIrZMxqJmAazDY0XFSgDmYPadXrNYAG60xgk2ct3kAX0ghvQA) looks like in AST:

image

The [documentation of the compiler](https://fsharp.github.io/fsharp-compiler-docs/) is searchable:
image

I hope this helps.

**Comment by ThisFunctionalTom** [comment]
Thanks for the links. 
I tried this already when I wrote Html2Feliz but wasn't very successful. 

Maybe I just need to invest more time in it. 😅

**Comment by nojaf** [comment]
If you ever want to pair on something like this, don't hesitate to reach out.
The same goes for you @WhiteBlackGoose.

**Comment by ThisFunctionalTom** [comment]
> If you ever want to pair on something like this, don't hesitate to reach out. The same goes for you @WhiteBlackGoose.

Oh, this is really nice of you. I would really like that. Maybe we could at least start converting Html2Feliz to Fantomas. 
And I have so many ideas for Myriad I would like to try.

**Comment by ThisFunctionalTom** [comment]
Sorry @WhiteBlackGoose for hijacking your case here but look what I did. Maybe it can help.

```fsharp
#r "nuget: Fantomas"

open System
open System.IO
open FSharp.Compiler
open FSharp.Compiler.Syntax
open FSharp.Compiler.Text
open FSharp.Compiler.Xml
open Fantomas

module Ast =
    let zeroRange = Range.Zero
    let zeroXml = PreXmlDoc.Empty
    let ident s = Ident(s, zeroRange)

    let longIdent s =
        SynType.LongIdent(LongIdentWithDots([ ident s ], []))

    let app funName args =
        SynType.App(longIdent funName, Some zeroRange, args |> List.map longIdent, [], Some zeroRange, true, zeroRange)

    let recordField key typeName =
        SynField([], false, Some(ident key), longIdent typeName, false, zeroXml, None, zeroRange)

    let record name recordFields =
        SynTypeDefn.SynTypeDefn(
            SynComponentInfo([], None, [], [ ident name ], zeroXml, false, None, zeroRange),
            SynTypeDefnRepr.Simple(SynTypeDefnSimpleRepr.Record(None, recordFields, zeroRange), zeroRange),
            [],
            None,
            zeroRange
        )

    let module' types = SynModuleDecl.Types(types, zeroRange)

    let file modules =
        ParsedInput.ImplFile(
            ParsedImplFileInput(
                "tmp.fsx",
                true,
                QualifiedNameOfFile(ident "Tmp"),
                [],
                [],
                [ SynModuleOrNamespace(
                      [],
                      false,
                      SynModuleOrNamespaceKind.AnonModule,
                      modules,
                      zeroXml,
                      [],
                      None,
                      zeroRange
                  ) ],
                (true, true)
            )
        )

module Fantomas =
    let format file =
        CodeFormatter.FormatASTAsync(file, "tmp.fsx", [], None, FormatConfig.FormatConfig.Default)
        |> Async.RunSynchronously

Ast.file [ Ast.module' [ Ast.record "Person" [ Ast.recordField "FirstName" "string" ] ] ]
|> Fantomas.format
```

And the output is
image

File: Json2Fs.Core/Library.fs [hunk 1]
```diff
@@ -114,17 +117,68 @@ let rec generateFSharpDTOsByJson (typename : string) (json : JsonElement) : Type
...
+                      zeroXml,
+                      [],
+                      None,
+                      zeroRange
+                  ) ],
+                (true, true)
+            )
+        )
+
+    CodeFormatter.FormatASTAsync(file, "tmp.fsx", [], None, FormatConfig.FormatConfig.Default)
```
Comment (WhiteBlackGoose) [inline Json2Fs.Core/Library.fs]: Okay, so I decided to get into it!

Just one minor thing, I think it'd be nicer to have a bit different formatting:

```suggestion
    let config = { FormatConfig.FormatConfig.Default with RecordMultilineFormatter = FormatConfig.MultilineFormatterType.NumberOfItems }

    CodeFormatter.FormatASTAsync(file, "tmp.fsx", [], None, config)
```

**Comment by WhiteBlackGoose** [comment]
@ThisFunctionalTom looks neat! Are you going to make a lib with this API?

File: Json2Fs.Core/Library.fs (hunk 1 above)
Comment (nojaf) [reply Json2Fs.Core/Library.fs]: I'll get to this later this week.

**Comment by nojaf** [comment]
@WhiteBlackGoose I've never used it but you could look at https://github.com/ionide/FsAst if you want to use a library.
I tend to create a helper module as @ThisFunctionalTom proposes in my own projects instead.

**Comment by ThisFunctionalTom** [comment]
> @ThisFunctionalTom looks neat! Are you going to make a lib with this API?

I am thinking about it. Maybe something like Feliz syntax would be possible. 🤔
But I will need some time for it. 😅

**Comment by WhiteBlackGoose** [comment]
Thanks, folks! (just that one minor thing to decide and I hope I'll merge)

**Comment by nojaf** [comment]
Circling back to this, this is a good use case for [Fabulous.AST](https://edgarfgp.github.io/Fabulous.AST/).

In short, since Fantomas v6 you can [generate code](https://fsprojects.github.io/fantomas/docs/end-users/GeneratingCode.html) using our internal AST (which is easier to reason with than the untyped F# AST). `Fabulous.AST` is a nice DSL on top of that created by @edgarfgp and really streamlines everything.

**Comment by WhiteBlackGoose** [comment]
Oh wow, looks like I forgot about this PR entirely, I'm sorry. Let me test it quickly

**Comment by WhiteBlackGoose** [comment]
Seems to be working, thanks

**Pull request merged**